### ISSUE TITLE
RunAsync crashing on upgrade on standalone cluster. Exit code:255

### ISSUE THREAD

**Issue by alexszym** [Azure/service-fabric-issues#1213]
I've encountered and issue during upgrade when attempting to upgrade a service from one without a RunAsync to one with a RunAsync consuming from a ReliableConcurrentQueue. The issue occurs on a standalone cluster (6.2.283.9494), I was not able to reproduce it on local environment. 

Here is the minimal code the issue is reproducible with. There is nothing populating the queue and the service is empty apart from the RunAsync:
```
        public async Task RunAsync(CancellationToken cancellationToken)
        {
            var eventsQueue = await StateManager.GetOrAddAsync<IReliableConcurrentQueue<QueuedEvent>>("Queue");

            try
            {
                while (true)
                {
                    cancellationToken.ThrowIfCancellationRequested();
                    try
                    {
                        using (var tx = StateManager.CreateTransaction())
                        {
                            var dequeuedEvent = await eventsQueue.TryDequeueAsync(tx, cancellationToken);
                            if (dequeuedEvent.HasValue)
                            {
                                await tx.CommitAsync();
                            }
                        }
                    }
                    catch (TimeoutException )
                    {
                    }
                    catch (TransactionFaultedException)
                    {
                    }
                    await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
                }
            }
            catch (OperationCanceledException )
            {
            }
        }
```
Here is the error I get on the deployed application package. As well as secondary replicas going down and not coming back up.

![error](https://user-images.githubusercontent.com/7223544/43652455-2a2f50ce-973d-11e8-9d9e-1cfa919df022.png)

After a while the primary replica ends up with reconfiguration being stuck:

![stuck](https://user-images.githubusercontent.com/7223544/43653084-ed275cd8-973e-11e8-923d-c155cff3d82d.png)

Here is the Exception that is being thrown after I've managed to dig down deeper into the issue:

```
FailFast: RunAsync failed due to an unhandled exception causing the host process to crash: System.ArgumentException: Returned reliable object of type Microsoft.ServiceFabric.Data.Collections.ReliableConcurrentQueue.ReliableConcurrentQueue`1[System.String] cannot be casted to requested type Microsoft.ServiceFabric.Data.Collections.IReliableConcurrentQueue`1[Akru.EventBus.ServiceFabric.QueuedEvent]
   at Microsoft.ServiceFabric.Data.ReliableStateManagerImpl.<Microsoft-ServiceFabric-Data-IReliableStateManager-GetOrAddAsync>d__26`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Data.ReliableStateManagerImpl.<Microsoft-ServiceFabric-Data-IReliableStateManager-GetOrAddAsync>d__28`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Akru.EventBus.ServiceFabric.DummyRunner.<RunAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DummyService.DummyService.<RunAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ServiceFabric.Services.Runtime.StatefulServiceReplicaAdapter.<ExecuteRunAsync>d__23.MoveNext()

   at System.Environment.FailFast(System.String, System.Exception)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
```

Here is the QueuedEvent for reference:
```
    [DataContract]
    public class QueuedEvent
    {
        public QueuedEvent(string serialisedData)
        {
            SerialisedData = serialisedData;
        }
        public QueuedEvent() { }

        [DataMember]
        public string SerialisedData { get; set; }
    }
```

The upgrade works fine if I switch to a queue of strings rather than queue of QueuedEvents. The issue seemed to be more reliably reproduced when primary replica is on the earlier upgrade domains than the secondary replicas.

Questions:
1. The error code was pretty obscure and took a while to find out what the actual issue is, is there anywhere a list of all the possible error codes and what they mean?
2. I don't understand why the service is attempting to deserialise the queue (which is empty) into a queue of strings, rather than queue of the QueuedEvent objects. Is there anything I'm doing wrong?

**Comment by masnider** [comment]
Can you share how the data is getting queued into the ReliableQueue ?

**Comment by yizhang82** [comment]
Based on this error:

> Returned reliable object of type Microsoft.ServiceFabric.Data.Collections.ReliableConcurrentQueue.ReliableConcurrentQueue`1[System.String] cannot be casted to requested type Microsoft.ServiceFabric.Data.Collections.IReliableConcurrentQueue`1[Akru.EventBus.ServiceFabric.QueuedEvent]

Are you changing the type of the queue?

**Comment by tyms** [comment]
Hi, is it possible that a ReliableConcurrentQueue<System.String> was previously created with the same name? Can you try calling Delete before upgrading again?

**Comment by alexszym** [comment]
Thanks for the responses.

I've isolated this example to a smallest chunk possible to try and determine the cause. The queue is not being populated at all. There is no Enqueueing. The steps I am using is deploying a fresh service with no RunAsync and upgrading to the a RunAsync version I've provided. Then deleting the whole application as cleanup.

There is nothing else creating the queue other than this and nothing creates ReliableConcurrentQueue<System.String> just the one with QueuedEvent, hence my bewilderment at this error.

**Comment by yizhang82** [comment]
I think it's the upgrading that is causing the problem.

Please keep in mind ReliableCollection is really a database wrapped in Dictionary interface - data is persisted as well as the type information. Most likely you have changed the type of the queue in the new version of the application. So in the older version you have the queue typed as `IReliableConcurrentQueue<QueuedEvent>`, and in the new version you have the queue typed as `IReliableConcurrentQueue<String>`. Because the data is persisted, we remember the type of the queue being `IReliableConcurrentQueue<QueuedEvent>` and would always return that in future iterations (since you can't change the queue type unless you delete it). This could be entirely accidental for you that the two queues may be unrelated but just so happens to have the same name. 

To mitigate this, if you don't need the previous data, you can use a new name for the new queue, and delete the old queue. If you care about the data, you can implement a copy routine as well. Another option is to keep the old type but change the type to whatever field you desire, while having a serializer that is compatible with both versions so that you can read old data and new data at the same time.

**Comment by alexszym** [comment]
The issue still occurs if I deploy the application under a different name with no RunAsync and nothing creating a queue with that name.

If the application has a different name I assume the queue would be an entirely different store on the cluster. I then follow the steps of upgrading the service it to the RunAsync version and I still encounter the issue.

There is no queue in that service before upgrade and after upgrade there is only the code that creates the queue with QueuedEvent type. There is no reference to IReliableConcurrentQueue<String>queue in the code with these steps.

Could this be a bug?

**Comment by alexszym** [comment]
Here is the entire code I'm using for the service. The only difference is before upgrade the RunAsync is commented out.

```
    internal sealed class DummyService : StatefulService
    {
        public DummyService(StatefulServiceContext context)
            : base(context)
        { }

        protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners()
        {
            return new ServiceReplicaListener[0];
        }

        protected override async Task RunAsync(CancellationToken cancellationToken)
        {
            // TODO: Replace the following sample code with your own logic 
            //       or remove this RunAsync override if it's not needed in your service.

            var eventsQueue = await StateManager.GetOrAddAsync<IReliableConcurrentQueue<string>>("Queue");

            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();

                using (var tx = StateManager.CreateTransaction())
                {
                    var result = await eventsQueue.TryDequeueAsync(tx, cancellationToken);
                    if (result.HasValue)
                    {
                        // If an exception is thrown before calling CommitAsync, the transaction aborts, all changes are 
                        // discarded, and nothing is saved to the secondary replicas.
                        await tx.CommitAsync();
                    }
                }

                await Task.Delay(TimeSpan.FromMilliseconds(100), cancellationToken);
            }
        }
    }
```

**Comment by yizhang82** [comment]
Does the type `Akru.EventBus.ServiceFabric.QueuedEvent` ring any bell to you? I don't think this is a bug as we wouldn't be able to "fabricate" (if you pardon the pun) that type out of thin air. My guess is that most likely this is a library that you use (A google search didn't hit anything) and that library happen to use the same name "Queue" (not surprisingly) as your code in the same service, hence the conflict.

**Comment by alexszym** [comment]
Yes it does ring a bell. As I've mentioned before I have a queue of QueuedEvent (which has the namespace Akru.EventBus.ServiceFabric.QueuedEvent).But the error I am getting on upgrade is that the queue is being deserialised into string type (and causing the application to crash), even though i have no reference to that in my code. Basically there is no queue of type string in the code at all.

Could that behaviour be a bug?

**Comment by yizhang82** [comment]
The error message indicates the other way around - that the code is trying to convert queue of string into queue of QueueEvent. 

Assuming the RunAsync code with string queue is what you had before upgrade, and the RunAsync code with QueueEvent queue is what you are trying to upgrade to, that is what is causing the issue. Since before upgrade you had the queue with string type, and we persist both the data and the type of the queue on disk. Next time when you are doing upgrade, we load up the previous type from disk and that is causing the crash. 

This is by design - as the on-disk representation of string and QueueEvent may be very different. From your point they are both strings under the hood, however the serialization can (and most likely will) choose to serialize the data differently. For example, most serializer will choose to encode the field ID as part of the serialized buffer for QueueEvent. And if they are exactly the same, service fabric doesn't know that because it has no underlying knowledge about the serialization format. For reasons exactly like these, we don't allow user to change the type of the queue or casting to a different type after the fact. We may choose to allow such scenarios in the future with a very explicit API (such as UnsafeCastTo, for example), and user would have to make sure the data format is indeed compatible. 

In summary:
* ReliableConcurrentQueue / Dictionary identity is strictly based on name. You can't have two of them with different type but of the same name. The first one gets created "wins". 
* The type of the queue is persisted on disk the first time it is created, and you can't change it after that, at least not today. The reason is for type safety. In the future we may allow you to do so explicitly. 
* Therefore, if you create a queue with type A but later try to reuse it as type B, that wouldn't work today.

Hope this helps.

**Comment by alexszym** [comment]
Hi @yizhang82 ,

Thanks for the explanation, that all makes sense. 

The problem I am describing here is an upgrade where the initial version doesn't have any reference to a queue of any type. 
Version 1.0 - My RunAsync is commented out. There is no queue creation code or any references to any queue. 
Version 2.0 - I uncomment the RunAsync, this should create a queue of QueuedEvent. 

With these steps I don't understand why the expected queue type is string. Nothing creates a queue of type string in my code. My expectation here is that on upgrade a queue of QueuedEvent type should be created, but I am seeing the ArgumentException instead, hence why I think this might be a bug.

I am able to reproduce this reliably by deploying the application under a different name (different application name means no initial state) and using the steps I've described.

**Comment by raunakpandya** [comment]
I think the confusion here is because in your last code snippet, you pasted queue of strings. 

var eventsQueue = await StateManager.GetOrAddAsync<IReliableConcurrentQueue<string>>("Queue");

**Comment by yizhang82** [comment]
Based on all the information above, my guess is that your v1 package might be stale and could contain a older version that had a string queue in it, even though you have updated the code to comment everything out.  I would double check the binary in your v1 package using ildasm/ilspy/refector (or whatever .NET decompiler you prefer) to look at the RunAsync code and see if there is any reference to a string queue. And really make sure the binaries are updated before deploy v1 again under a new application name.

**Comment by alexszym** [comment]
@raunakpandya Thanks for that. I was copying it from a solution with the fix. I have updated the comment to reflect the actual code.

@yizhang82 Thanks, I'll give it a try.

**Comment by alexszym** [comment]
I've decompiled both the version of the code before the upgrade and after and none of them had a reference to a string queue, so still not sure what caused this.

We've upgraded to a 6.3.162.9494 version of the standalone cluster since then and I'm no longer able to reproduce the behaviour, so closing the issue.
